timestamp-client/src/main/java/com/palantir/timestamp/RequestBatchingTimestampService.java

```diff
@@ -73,211 +55,77 @@ public boolean isInitialized() {
-
-            valueToReturnNext = new AtomicLong(-1L); // Dummy value. Will be populated.
-            failed = false;
-        }
-
-        // The following methods must only be called by the thread that created this batch.
-        public void populate(TimestampRange range) {
-            this.endInclusive = range.getUpperBound();
-            this.valueToReturnNext.set(range.getLowerBound());
-        }
-        public int getRequestCountAndSetInvalid() {
-            return requestCount.getAndSet(Integer.MIN_VALUE);
-        }
-        public void fail() {
-            failed = true;
-        }
-        public void becomeReadable() {
-            populationLatch.countDown();
-        }
-        // End creator-only threads.
-
-        public boolean isFailed() {
-            return failed;
-        }
-
-        /**
-         * Attempts to join a timestamp request batch.
-         *
-         * @return true if we are included in the batch and false otherwise
-         */
-        public boolean joinBatch() {
-            if (requestCount.get() < 0) {
-                return false;
+    @VisibleForTesting
+    static Consumer<List<BatchElement<Integer, TimestampRange>>> consumer(TimestampService delegate) {
+        return batch -> {
+            long totalTimestamps = batch.stream().mapToLong(BatchElement::argument).reduce(0, Math::addExact);
+            long startInclusive = 0;
+            long endExclusive = 0;
+            for (BatchElement<Integer, TimestampRange> element : batch) {
+                int timestampsRequired = element.argument();
+                if (element.argument() == 0) {
+                    element.result().setException(new IllegalArgumentException("Cannot ask for 0 timestamps"));
+                } else if (element.argument() <= endExclusive - startInclusive) {
+                    element.result().set(createExclusiveRange(startInclusive, startInclusive + timestampsRequired));
+                    startInclusive += timestampsRequired;
+                    totalTimestamps -= timestampsRequired;
+                } else {
+                    TimestampRange requested = getFreshTimestampsFromDelegate(
+                            delegate, Ints.saturatedCast(totalTimestamps));
+                    startInclusive = requested.getLowerBound();
+                    endExclusive = Math.addExact(requested.getUpperBound(), 1);
+                    int toTake = Math.min(Ints.checkedCast(endExclusive - startInclusive), timestampsRequired);
+                    element.result().set(createExclusiveRange(startInclusive, startInclusive + toTake));
+                    startInclusive += toTake;
+                    totalTimestamps -= timestampsRequired;
+                }
             }
-            int val = requestCount.incrementAndGet();
-            return isRequestCountValid(val);
-        }
-
-        public void awaitPopulation() throws InterruptedException {
-            populationLatch.await();
-        }
+        };
+    }
```

Comment:
Nice (and this also adds support for batching `getFreshTimestamps`). The multiple case looks correct to me as well (considering that you only need to return at-least-1 timestamp).

It seems a bit funky to say "cannot ask for 0 timestamps" after making the request(s), but not a major concern.